File: 2074-reverse-nodes-in-even-length-groups/2074-reverse-nodes-in-even-length-groups.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int reverse(ListNode* last,ListNode* headOfGroup,int n){
        ListNode* prev = nullptr;
        ListNode* curr = headOfGroup;
        ListNode* nextNode = curr->next;
        int count =0;

        while(curr !=nullptr && n>0){
            nextNode = curr->next;

            curr->next = prev;
            prev = curr;
            curr = nextNode;
            n--;
            count++;
        }
        headOfGroup->next = nextNode;
        last->next = prev;
        return count;
    }
    ListNode* reverseEvenLengthGroups(ListNode* head) {
        int gn =0;
        int elementsInPrevGroup =0;
        ListNode* curr = head;
        ListNode* lastOfPrevGroup = nullptr;
        ListNode* lastOfEvenGroup = nullptr;

        while(curr!=nullptr){
            gn++;
            if(gn%2==0){
                elementsInPrevGroup = reverse(lastOfPrevGroup,curr,gn);
                lastOfEvenGroup = curr;
                curr = curr->next;
            }else{
                int k = gn;
                elementsInPrevGroup =0;
                while(curr!=nullptr && k>0){
                    lastOfPrevGroup = curr;
                    curr = curr->next;
                    k--;
                    elementsInPrevGroup++;
                }
            }
        }

        if(gn % 2 ==1 && elementsInPrevGroup %2 ==0){
            reverse(lastOfEvenGroup,lastOfEvenGroup->next,elementsInPrevGroup);
        }else if(gn % 2 ==0 && elementsInPrevGroup %2 ==1){
            reverse(lastOfPrevGroup,lastOfPrevGroup->next,elementsInPrevGroup);
        }
        return head;
    }
};

File: 2074-reverse-nodes-in-even-length-groups/2074-reverse-nodes-in-even-length-groups.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverse(self, last, headOfGroup, n):
        prev = None
        curr = headOfGroup
        nextNode = curr.next
        count = 0

        while curr != None and n > 0:
            nextNode = curr.next

            curr.next = prev
            prev = curr
            curr = nextNode
            n -= 1
            count += 1
        headOfGroup.next = nextNode
        last.next = prev
        return count

    def reverseEvenLengthGroups(self, head: ListNode | None) -> ListNode | None:
        gn = 0
        elementsInPrevGroup = 0
        curr = head
        lastOfPrevGroup = None
        lastOfEvenGroup = None

        while curr != None:
            gn += 1
            if gn % 2 == 0:
                elementsInPrevGroup = self.reverse(lastOfPrevGroup, curr, gn)
                lastOfEvenGroup = curr
                curr = curr.next
            else:
                k = gn
                elementsInPrevgroup = 0
                while curr != None and k > 0:
                    lastOfPrevGroup = curr
                    curr = curr.next
                    k -= 1
                    elementsInPrevGroup += 1

        if gn % 2 == 1 and elementsInPrevGroup % 2 == 0:
            self.reverse(lastOfEvenGroup, lastOfEvenGroup.next, elementsInPrevGroup)
        elif gn % 2 == 0 and elementsInPrevGroup % 2 == 1:
            self.reverse(lastOfPrevGroup, lastOfPrevGroup.next, elementsInPrevGroup)
        return head
